### PR TITLE
ES similarity_search_with_score() and metadata filter

### PR DESCRIPTION
**Pull request by Zzz233**
Add similarity_search_with_score() to ElasticVectorSearch, add metadata filter to both similarity_search() and similarity_search_with_score()

### CODE REVIEW

File: langchain/vectorstores/elastic_vector_search.py [hunk 1]
```diff
@@ -199,14 +208,42 @@ def similarity_search(
...
+            k: Number of Documents to return. Defaults to 4.
+        Returns:
+            List of Documents most similar to the query.
+        """
+        embedding = self.embedding.embed_query(query)
+        script_query = _default_script_query(embedding, filter)
+        response = self.client.search(
+            index=self.index_name, query=script_query, size=k)
+        hits = [hit for hit in response["hits"]["hits"]]
+        documents = [
```
Comment (dev2049) [inline langchain/vectorstores/elastic_vector_search.py:238]: nit: could we call this something like `docs_and_scores`

File: langchain/vectorstores/elastic_vector_search.py [hunk 2]
```diff
@@ -187,7 +192,11 @@ def add_texts(
         return ids
 
     def similarity_search(
```
Comment (dev2049) [inline langchain/vectorstores/elastic_vector_search.py:194]: should this just call `similarity_search_with_score` and drop the scores

File: langchain/vectorstores/elastic_vector_search.py [hunk 3]
```diff
@@ -20,10 +20,15 @@ def _default_text_mapping(dim: int) -> Dict:
     }
 
 
-def _default_script_query(query_vector: List[float]) -> Dict:
+def _default_script_query(query_vector: List[float], filter: dict) -> Dict:
```
Comment (dev2049) [inline langchain/vectorstores/elastic_vector_search.py:23]: should typing be `filter: Optional[dict]`, given how it's called

File: langchain/vectorstores/elastic_vector_search.py [hunk 4]
```diff
@@ -20,10 +20,15 @@ def _default_text_mapping(dim: int) -> Dict:
...
+    if filter:
+        (key, value), = filter.items()
```
Comment (dev2049) [inline langchain/vectorstores/elastic_vector_search.py:25]: could there be a valid filter with more than one key/val pair? if not could we explicitly check filter length

**Review by dev2049 — COMMENTED** [review]
could we add simple integration test for filter and similarity with score

File: langchain/vectorstores/elastic_vector_search.py (hunk 4 above)
Comment (Zzz233) [reply langchain/vectorstores/elastic_vector_search.py:25]: github_1

Pdfloader (PyMuPDF) extracts metadata including source, file path, page number, total pages, format, title, author, keywords, creator, producer, creation date, mod date, and trapped.

I find filtering by file path is very useful. I am a newbie, Perhaps you could demonstrate how to compile that for me.

**Pull request merged**